pkg/controller/certificates/certificate_controller.go

```diff
@@ -186,15 +179,21 @@ func (cc *CertificateController) maybeSignCertificate(key string) error {
 	defer func() {
 		glog.V(4).Infof("Finished syncing certificate request %q (%v)", key, time.Now().Sub(startTime))
 	}()
-	obj, exists, err := cc.csrStore.Store.GetByKey(key)
+	csr, err := cc.csrLister.Get(key)
+	if errors.IsNotFound(err) {
+		glog.V(3).Infof("csr has been deleted: %v", key)
+		return nil
+	}
 	if err != nil {
 		return err
 	}
-	if !exists {
-		glog.V(3).Infof("csr has been deleted: %v", key)
-		return nil
+
+	// need to operate on a copy so we don't mutate the csr in the shared cache
+	csrCopy, err := api.Scheme.DeepCopy(csr)
```

Comment:
It's not a simple short-circuit. If the approver is nil, there's still a chance that the csr is approved. So you'd need to deal with potential copies in both the non-nil approver block and the "need to sign" block.